Makefile

```diff
@@ -134,3 +134,125 @@ clean:
 
 expression: format clippy
 	LOG_LEVEL=ERROR RUST_BACKTRACE=1 cargo test --features "${ENABLE_FEATURES}" "coprocessor::dag::expr" -- --nocapture
+
+
+
+# The below x- targets are temporary, for experimenting with new profiles,
+# specifically in pursuit of compile time speedups.
+#
+# re https://github.com/tikv/tikv/issues/4189
+#
+# The idea here is that there are more "profiles" than just "dev" and "release".
+# In particular, there is an optimized dev profile, here "dev-opt". The below
+# profiles are intentionally named differently from the stock cargo 'dev'
+# and 'release' profiles to avoid confusion, but eventually we might expect
+# e.g. dev-opt to become the 'release' and 'bench' profiles, and 'dev-opt'
+# to become the 'dev' and perhaps 'test' profiles; with the _real_ release
+# profile being created with a config file.
+#
+# They can be invoked as:
+#
+#     $ make x-build-dev-nopt      # An unoptimized build
+#                                  #   (fast build / slow run)
+#     $ make x-build-dev-opt       # A mostly-optimized dev profile
+#                                  #   (slower build / faster run)
+#     $ make x-build-prod          # A release build
+#                                  #   (slowest build / fastest run)
+#     $ make x-bench               # Run benches mostly-optimized
+#                                  #   (slower build / faster run)
+#     $ make x-test                # Run tests unoptimized
+#                                  #   (fast build / slow run)
+#
+# The first three have aliases:
+#
+#     $ make x-build
+#     $ make x-build-opt
+#     $ make x-release
+#
+# The below rules all rely on using a .cargo/config file to override various
+# profiles. Within those config files we'll experiment with compile-time
+# optimizations which can't be done with Cargo.toml alone.
+#
+# Eventually, we'll merge as much of the configs into Cargo.toml as possible,
+# and merge the below commands into the rest of the makefile.
+#
+# None of the build profiles has debuginfo on by default because it increases
+# the build time by ~20%. The easiest way to build with debuginfo is by setting
+# the DEBUG makefile variable,
+#
+#     $ make x-build DEBUG=1
+#
+# To pass extra arguments to cargo you can set CARGO_ARGS,
+#
+#     $ make x-build CARGO_ARGS="--all"
+
+DEV_OPT_CONFIG=etc/cargo.config.dev-opt
+DEV_NOPT_CONFIG=etc/cargo.config.dev-nopt
+PROD_CONFIG=etc/cargo.config.prod
+TEST_CONFIG=etc/cargo.config.test
+BENCH_CONFIG=etc/cargo.config.bench
+
+ifneq ($(DEBUG),)
+export X_DEBUG=${DEBUG}
+endif
+
+export X_CARGO_ARGS:=${CARGO_ARGS}
+
+x-build-dev-opt: export X_CARGO_CMD=build
+x-build-dev-opt: export X_CARGO_FEATURES=${ENABLE_FEATURES}
+x-build-dev-opt: export X_CARGO_RELEASE=1
+x-build-dev-opt: export X_CARGO_CONFIG_FILE=${DEV_OPT_CONFIG}
+x-build-dev-opt:
+	bash scripts/run-cargo.sh
```

Comment:
How about `env X_CARGO_CMD=build X_CARGO_RELEASE=1 bash scripts/run-cargo.sh`?

Reply:
All four of the previous line's env vars could be written that way, but I think it is less readable.

If your concern is about "leaking" the environment, the `target: env-var` syntax is local to the makefile target. In effect it causes make to export them similarly to what you wrote.